File: app/api/routes/library.py
from fastapi import APIRouter
from fastapi.encoders import jsonable_encoder
from app.db.database import SessionDep
from app.models.library import Library, LibraryVisibility, LibraryWithWordsId, LibraryDetail
from app.api.dependencies import get_current_user
from app.db.library import get_libraries_, create_library_, remove_library_, get_library_by_id, toggle_user_library_select_, get_library_details_, update_library_details_, update_user_library_select_priority_
from app.models.user import User
from app.api.api_responses import success_response, forbidden_response, not_found_response
from fastapi import Depends
from app.db.word import get_words_by_ids
from app.models.user_library_select import UpdatePriorityRequest

router = APIRouter(prefix="/api/library", tags=["library"])

@router.get("/get_libraries")
async def get_libraries(session: SessionDep, user_in: User = Depends(get_current_user)):
    return success_response(data=jsonable_encoder(get_libraries_(session, user_in)))

@router.post("/create_library")
async def create_library(session: SessionDep, library_in: Library, user_in: User = Depends(get_current_user)):
    library_in.creator_id = user_in.id
    library_in.visibility = LibraryVisibility.PRIVATE
    create_library_(session, library_in)
    return success_response(message=f"创建词库{library_in.name}成功！")

@router.delete("/remove_library/{library_id}")
async def remove_library(session: SessionDep, library_id: int, user_in: User = Depends(get_current_user)):
    library = get_library_by_id(session, library_id)
    if library is None:
        return not_found_response(message="词库不存在！")
    if library.creator_id != user_in.id:
        return forbidden_response(message="无权限删除词库！")
    remove_library_(session, library)
    return success_response(message=f"删除词库{library.name}成功！")

@router.post("/toggle_user_library_select/{library_id}")
async def toggle_user_library_select(session: SessionDep, library_id: int, user_in: User = Depends(get_current_user)):
    library = get_library_by_id(session, library_id)
    if library is None:
        return not_found_response(message="词库不存在！")
    selected = toggle_user_library_select_(session, user_in, library)
    return success_response(message=f"词库{library.name}{'成功选中' if selected else '取消选中'}！")

@router.put("/batch_update_library_priorities")
async def batch_update_library_priorities(
    session: SessionDep, 
    priorities: UpdatePriorityRequest,
    user_in: User = Depends(get_current_user)
):
    for item in priorities.priorities:
        library = get_library_by_id(session, item.library_id)
        if library:
            update_user_library_select_priority_(session, user_in, library, item.priority)
    return success_response(message="词库优先级更新成功！")

@router.get("/get_library_details/{library_id}")
async def get_library_details(session: SessionDep, library_id: int, user_in: User = Depends(get_current_user)):
    library = get_library_by_id(session, library_id)
    if library is None:
        return not_found_response(message="词库不存在！")
    return success_response(data=jsonable_encoder(get_library_details_(session, library_id)))

@router.put("/update_library/{library_id}")
async def update_library(session: SessionDep, library_id: int, library_in: LibraryWithWordsId, user_in: User = Depends(get_current_user)):
    library_in.id = library_id
    if get_library_by_id(session, library_id) is None:
        return not_found_response(message="词库不存在！")
    
    # 构建 LibraryDetail 对象
    library_detail = LibraryDetail.model_validate(library_in, from_attributes=True)
    library_detail.words = get_words_by_ids(session, library_in.words_id)
    
    update_library_details_(session, library_detail)
    return success_response(message=f"更新词库{library_in.name}成功！")

File: src/stores/library.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import request from '@/utils/request'

export interface Library {
  id: number
  name: string
  description: string
  creator_id: number
  visibility: 'public' | 'private'
  created_at: string
  word_count: number
  updated_at: string
  selected: boolean
  priority: number
}

export interface Word {
  id: number
  text: string
  meaning: string
  phonetic?: string
}

export interface LibraryDetails extends Library {
  words: Word[]
}

export interface BatchRecognizeResult {
  recognizedCount: number
  unrecognizedCount: number
  recognizedWords: Word[]
  unrecognizedWords: string[]
}

export const useLibraryStore = defineStore('library', () => {
  const libraries = ref<Library[]>([])

  async function fetchLibraries() {
    try {
      const res = await request.get('/api/library/get_libraries')
      libraries.value = res.data.data as Library[]

      // 分离已选择和未选择的词库
      const selected = libraries.value.filter((lib) => lib.selected)
      const unselected = libraries.value.filter((lib) => !lib.selected)

      // 已选择的按 priority 排序，未选择的按 id 排序
      selected.sort((a, b) => a.priority - b.priority)
      unselected.sort((a, b) => a.id - b.id)

      // 合并：已选择的在前，未选择的在后
      libraries.value = [...selected, ...unselected]
    } catch (error) {
      console.error('获取词库失败:', error)
      libraries.value = []
    }
  }

  async function getLibraries(forceRefresh = false) {
    if (forceRefresh || !libraries.value.length) {
      await fetchLibraries()
    } else {
      fetchLibraries()
    }
    return libraries.value
  }

  async function createLibrary(name: string, description: string) {
    const res = await request.post('/api/library/create_library', {
      name: name,
      description: description,
    })
    if (res.data.success) {
      await getLibraries(true)
    }
    return { success: res.data.success, message: res.data.message }
  }

  async function toggleLibrary(libraryId: number) {
    const res = await request.post(`/api/library/toggle_user_library_select/${libraryId}`)
    if (res.data.success) {
      await getLibraries(true)
    }
    return { success: res.data.success, message: res.data.message }
  }

  async function getLibraryDetails(libraryId: number): Promise<LibraryDetails> {
    const res = await request.get(`/api/library/get_library_details/${libraryId}`)
    if (res.data.success) {
      return res.data.data as LibraryDetails
    } else {
      throw new Error(res.data.message || '获取词库详情失败')
    }
  }

  async function searchWords(query: string): Promise<Word[]> {
    try {
      const res = await request.get('/api/word/search', {
        params: { q: query },
      })
      if (res.data.success) {
        return res.data.data as Word[]
      } else {
        return []
      }
    } catch (error) {
      console.error('搜索单词失败:', error)
      return []
    }
  }

  async function batchRecognizeWords(words: string[]): Promise<BatchRecognizeResult> {
    const res = await request.post('/api/word/batch_recognize', { words })
    if (res.data.success) {
      return res.data.data as BatchRecognizeResult
    } else {
      throw new Error(res.data.message || '批量识别单词失败')
    }
  }

  async function updateLibrary(
    libraryId: number,
    data: { name: string; description: string; words_id: number[] },
  ) {
    const res = await request.put(`/api/library/update_library/${libraryId}`, data)
    if (res.data.success) {
      await getLibraries(true)
      return true
    } else {
      return false
    }
  }

  async function removeLibrary(libraryId: number) {
    const res = await request.delete(`/api/library/remove_library/${libraryId}`)
    if (res.data.success) {
      await getLibraries(true)
    }
    return {
      success: res.data.success,
      message: res.data.message || (res.data.success ? '删除词库成功' : '删除词库失败'),
    }
  }

  async function batchUpdateLibraryPriorities(
    priorities: Array<{ library_id: number; priority: number }>,
  ) {
    const res = await request.put('/api/library/batch_update_library_priorities', { priorities })
    return { success: res.data.success, message: res.data.message }
  }

  return {
    libraries,
    fetchLibraries,
    getLibraries,
    createLibrary,
    toggleLibrary,
    getLibraryDetails,
    searchWords,
    batchRecognizeWords,
    updateLibrary,
    removeLibrary,
    batchUpdateLibraryPriorities,
  }
})
